### PR TITLE
lums/sc 37627/indexivfflat Group schema

### PR DESCRIPTION
**Pull request by lums658**
This PR creates a group class `ivf_flat_index_group` for the `ivf_index` class that abstracts the `TileDB::Group` representing an IVF Flat Index.  It also rolls up preceding PRs 131, 149, 153, and 154.  The primary changes for this PR were to add the files `ivf_flat_group.h` and `ivf_flat_metadata.h` and to incorporate the revised access to group information into the `ivf_flat_index` class contained in `ivf_flat_index.h`.  Also added are unit tests for `ivf_flat_group` and `ivf_flat_metadata`.

The group and metadata classes to a fair amount of verification and testing of the data available in the TileDB::group and its associated metadata.

### Representing information about the `TileDB::Group`

The `TileDB::Group`  always contains the following arrays, which are the minimum necessary to represent an IVF Flat Index:
```text
  partition_centroids
  partition_indexes
  shuffled_vector_ids
  shuffled_vectors
```
Note that there is a level of indirection between `ivf_flat_index_group` and these names.  The names may change as TileDB-Vector-Search evolves.  A version number is used to distinguish different versions of the group data.

Other arrays that the Group may contain are:
```c++
  input_vectors
  external_ids
  updates
```

The Group contains another Group 
```c++
temp_data
```

Finally, the group contains information about attribute filters, and whether or not it supports time traveling, with the tags
```c++
default_attr_filters
support_timetravel
```
This information is captured by `ivf_flat_index_group`.  A `uniquie_ptr` to an `ivf_flat_index_group` is now part of the `ivf_flat_index` class.

### Index Arrays

The index represented by an `ivf_flat_index_group` support insertions, deletions, and consolidation.  As a result, the size of the matrix or vector being stored in one of the group arrays will change over time.  To handle this, the domain for each of the arrays is set to  go from zero to `std::numeric_limits<type>::max() - 1`.  To specify how much of each such array is active and should be read in, we keep information about the index dimension, its overall size (number of vectors and ids), and its number of partitions (equal to number of centroids and one less than the index size).  If time traveling is enabled, this information is updated on each consolidation.  The `ivf_flat_index_group` class has methods for getting the most recent value of each of these quantities, as well as to append a new value.  Access to arbitrary timestamps will be added as part of implementing time traveling.

### Metadata
The group also contains metadata with further information about the Group:
```c++
      dataset_type
      index_type
      storage_version
      dtype
      feature_type
      id_type
      indices_type
      base_sizes
      ingestion_timestamps
      partition_history
      temp_size
      index_kind
      dimension
      feature_datatype
      id_datatype
      px_datatype
```
These metadata are captured by `ivf_flat_index_metadata`.  An `ivf_flat_index_metadata` object is a member of the `ivf_flat_index_group` class.

These metadata are a superset of what is currently defined in the the Python API.  The current `ivf_flat_index_group` class and `ivf_flat_index_metadata` class should be able to read an index group created by the Python API.  However, due to the lack of a set of meaningful index group arrays, this conjecture has not been tested.

### Time traveling

Note that time traveling is currently not fully implemented in this PR, but will be added as a PR in the very near future.  The necessary infrastructure to support time traveling is part of `ivf_flat_index_group` and `ivf_flat_index_metadata`.  Time traveling is intimately related to insertions and deletions to the index.  Partial support for that has been added as well.  A PR in the near future will add this feature.

### Unit testing

The `ivf_flat_index_group` and `ivf_flat_index_metadata` classes have associated unit tests.  Each class has a compare method to compare `self` with another object.  These checks have been incorporated into the `compare` function of `ivf_flat_index`; tests of `ivf_flat_index` will now implicitly also check `ivf_flat_index_group` and `ivf_flat_index_metadata`.

The primary changes to `ivf_flat_index` to use `ivf_flat_index_group` are via calls to the group to obtain information about the sizes of arrays (matrix and vector) that will be read in.

The API for `ivf_flat_index` did not itself change.  The CLI programs were lightly tested/verified for functionality and correctness, but they were not benchmarked.

The following unit tests are currently known to fail but were not fixed in the interest of getting this PR out:
```
  unit_vamana_test
  unit_fixed_min_heap_test
  unit_slicing_test
```
`unit_vamana_test` will be updated along with future PRs having to do with `vamana_index`.  The others are lower level and do not impact any active code.  They will be pruned and/or fixed in a future PR.

### Misc

Also added was the small fix to `tdbPartitionedMatrix` that was identified in PR #150.

Compilation and testing was done on MacOS 14.1.2 using Apple clang.  It has not been compiled or tested with g++.

### CODE REVIEW

File: src/CMakeLists.txt
Comment (jparismorgan) [inline src/CMakeLists.txt:163]: I see we're also checking in `src/cmake/HighFive.cmake` - should this be uncommented then? Or should we not check in `src/cmake/HighFive.cmake`?

File: src/include/api/feature_vector_array.h
Comment (jparismorgan) [inline src/include/api/feature_vector_array.h:241]: (Curious) What is the strategy on when to use `[nodiscard]]` vs not?

File: src/include/api/feature_vector_array.h
Comment (jparismorgan) [inline src/include/api/feature_vector_array.h:320]: (Minor) Remove?

File: src/include/api/flat_l2_index.h
Comment (jparismorgan) [inline src/include/api/flat_l2_index.h:115]: (Minor) Could this and `num_vectors()` be `const`?

File: src/include/detail/graph/diskann.h
Comment (jparismorgan) [inline src/include/detail/graph/diskann.h:91]: (Optional) Maybe a unit test to confirm?

File: src/include/detail/graph/vamana.h
Comment (jparismorgan) [inline src/include/detail/graph/vamana.h:990]: (Unknown) Should the `return false;` be commented back in?

File: src/include/test/CMakeLists.txt
Comment (jparismorgan) [inline src/include/test/CMakeLists.txt:116]: Is this (and the others below) commented out because it's failing? If not, could we comment back in?

File: src/src/CMakeLists.txt
Comment (jparismorgan) [inline src/src/CMakeLists.txt:138]: (Optional) Should we just remove the `TILEDB_VS_ENABLE_BLAS` and the if/else condition?

File: src/src/flat/CMakeLists.txt
Comment (jparismorgan) [inline src/src/flat/CMakeLists.txt:29]: Should these be commented back in? Or for a later PR?

**Review by jparismorgan — APPROVED** [review]
I read everything, but I might have missed things, so I would suggest to wait for another review before landing.

**Comment by lums658** [comment]
Superseded by #221.

**Pull request closed without merging**
